pkg/controller/utils/taints.go

```diff
@@ -0,0 +1,46 @@
+package utils
+
+import (
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func IdentifierForTaint(taint *corev1.Taint) hivev1.TaintIdentifier {
+	// Let a nil `taint` panic
+	return hivev1.TaintIdentifier{
+		Key:    taint.Key,
+		Effect: taint.Effect,
+	}
+}
+
+// GetUniqueTaints collapses the duplicates from the list of taints provided before returning them.
+// Key+Effect are unique identifiers, and the first Value of the taint entry encountered will be preserved.
+func GetUniqueTaints(taints *[]corev1.Taint) *[]corev1.Taint {
+	// uniqueTaints would be a map of key + effect to taint
+	uniqueTaints := make(map[hivev1.TaintIdentifier]corev1.Taint)
+	for _, taint := range *taints {
+		// If value is already present (we have encountered a duplicate), skip.
+		// This means that when we encounter duplicates, the value of the first entry encountered will be preserved.
+		if !TaintExists(uniqueTaints, &taint) {
+			uniqueTaints[IdentifierForTaint(&taint)] = taint
+		}
+	}
+	return ListFromTaintMap(&uniqueTaints)
+}
+
+// ListFromTaintMap simply compiles a list of taints from the corresponding map
+func ListFromTaintMap(taintMap *map[hivev1.TaintIdentifier]corev1.Taint) *[]corev1.Taint {
```

Comment:
So I thought about it, but really, we cannot guarantee any order as long as a map is involved. If fact, even multiple append/updates will result in k8s returning the list with values added as they encounter the request - so the list itself could also be in different order.
I have taken care to reword so we say first entry "encountered" will be preserved, not the topmost.
As far as the order changing arbitrarily, we can at the least guarantee CR will only be updated when necessary.

Reply:
We can guarantee ordering as long as we _only use maps for dup checking_. Roughly:

```
// Used to detect dups
uniques := map[TaintIdentifier]Taint{}
// Built up in order, first by generated, then by remote
taints := []Taint{}
// MachinePool takes precedence, so loop it first
for taint in gMS.Taints:
  id4t := IdentifierForTaint(taint)
  // We shouldn't have to consult taintsToDelete in this loop
  // because they'll be absent from gMS.Taints anyway.
  if _, ok := uniques[id4t]; ok:
    // Dup being squashed out
    objectModified = true
  else:
    // Note that this may be adding a new taint, but we don't
    // know that yet for purposes of objectModified; we'll figure
    // it out later.
    taints = append(taints, taint)
    uniques[id4t] = taint

// Now process remotes, differently.
// This set will be used to decide if the above loop *added* taints.
origRemotes := Set[TaintIdentifier]{}
for taint in rMS.Taints:
  id4t := IdentifierForTaint(taint)
  origRemotes.Add(id4t)
  if dup, ok := uniques[id4t]; ok:
    // This may be a modification (values differ) *or* a no-op
    // (values are the same)
    if dup.Value != taint.Value:
      objectModified = true
  elif _, ok := taintsToDelete[id4t]; ok:
    // Owned taint removed from MP
    objectModified = true
  else:
    taints = append(taints, taint)
    // Ensure dups that are only in the rMS get squashed too
    uniques[id4t] = taint

// At this point, if uniques contains any keys that weren't in the
// original rMS, it means we added taints from the MP.
for k := range uniques:
  if !origRemotes.Has(k):
    objectModified = true
    break
```